File: backend/app/api/routes/ai.py
"""AI Health Intelligence API — educational assistant only."""

from fastapi import APIRouter, Depends, HTTPException, Query

from app.ai.models import (
    AiBloodworkReportRequest,
    AiBloodworkReportResult,
    AiChatRequest,
    AiChatResponse,
    AiCycleReportRequest,
    AiCycleReportResult,
    AiInsightsRequest,
    AiInsightsResult,
    AiTimelineRequest,
    AiTimelineResult,
)
from app.core.auth import AuthenticatedUser, get_current_user
from app.services import ai_service

router = APIRouter(prefix="/ai", tags=["ai"])


@router.post("/bloodwork-report", response_model=AiBloodworkReportResult)
async def bloodwork_report(
    body: AiBloodworkReportRequest,
    user: AuthenticatedUser = Depends(get_current_user),
) -> AiBloodworkReportResult:
    """
    Generate educational bloodwork report from structured data.
    AI explains — does NOT diagnose or calculate risk scores.
    """
    try:
        return await ai_service.generate_bloodwork_report(user.user_id, body)
    except ValueError as exc:
        raise HTTPException(status_code=429, detail=str(exc)) from exc


@router.post("/cycle-report", response_model=AiCycleReportResult)
async def cycle_report(
    body: AiCycleReportRequest,
    user: AuthenticatedUser = Depends(get_current_user),
) -> AiCycleReportResult:
    """Generate educational cycle report explaining rule-based risk scores."""
    try:
        return await ai_service.generate_cycle_report(user.user_id, body)
    except ValueError as exc:
        raise HTTPException(status_code=429, detail=str(exc)) from exc


@router.post("/timeline", response_model=AiTimelineResult)
async def health_timeline(
    body: AiTimelineRequest,
    user: AuthenticatedUser = Depends(get_current_user),
) -> AiTimelineResult:
    """Generate chronological educational health timeline."""
    try:
        return await ai_service.generate_timeline(user.user_id, body)
    except ValueError as exc:
        raise HTTPException(status_code=429, detail=str(exc)) from exc


@router.post("/insights", response_model=AiInsightsResult)
async def health_insights(
    body: AiInsightsRequest,
    user: AuthenticatedUser = Depends(get_current_user),
) -> AiInsightsResult:
    """Generate factual educational insights from trends."""
    try:
        return await ai_service.generate_insights(user.user_id, body)
    except ValueError as exc:
        raise HTTPException(status_code=429, detail=str(exc)) from exc


@router.post("/chat", response_model=AiChatResponse)
async def ai_chat(
    body: AiChatRequest,
    user: AuthenticatedUser = Depends(get_current_user),
) -> AiChatResponse:
    """
    Educational chat — structured context only, no raw prompt passthrough.
    Does not provide medical advice or prescribe compounds.
    """
    try:
        return await ai_service.chat(user.user_id, body)
    except ValueError as exc:
        raise HTTPException(status_code=429, detail=str(exc)) from exc


@router.get("/chat/history")
async def chat_history(
    user: AuthenticatedUser = Depends(get_current_user),
    limit: int = Query(20, ge=1, le=100),
) -> list[dict]:
    """Retrieve recent chat messages for the current user."""
    return await ai_service.get_chat_history(user.user_id, limit)

File: frontend/services/ai.ts
import { authenticatedFetch } from "@/lib/api/authenticatedFetch";
import type {
  AiBloodworkReportRequest,
  AiBloodworkReportResult,
  AiChatRequest,
  AiChatResponse,
  AiCycleReportRequest,
  AiCycleReportResult,
  AiInsightsRequest,
  AiInsightsResult,
  AiTimelineRequest,
  AiTimelineResult,
  ChatHistoryMessage,
} from "@/types/ai";

const BASE = "/api/v1/ai";

export async function generateBloodworkReport(
  body: AiBloodworkReportRequest,
  accessToken: string | null | undefined
): Promise<AiBloodworkReportResult> {
  return authenticatedFetch<AiBloodworkReportResult>(`${BASE}/bloodwork-report`, {
    method: "POST",
    body,
    accessToken,
  });
}

export async function generateCycleReport(
  body: AiCycleReportRequest,
  accessToken: string | null | undefined
): Promise<AiCycleReportResult> {
  return authenticatedFetch<AiCycleReportResult>(`${BASE}/cycle-report`, {
    method: "POST",
    body,
    accessToken,
  });
}

export async function generateTimeline(
  body: AiTimelineRequest,
  accessToken: string | null | undefined
): Promise<AiTimelineResult> {
  return authenticatedFetch<AiTimelineResult>(`${BASE}/timeline`, {
    method: "POST",
    body,
    accessToken,
  });
}

export async function generateInsights(
  body: AiInsightsRequest,
  accessToken: string | null | undefined
): Promise<AiInsightsResult> {
  return authenticatedFetch<AiInsightsResult>(`${BASE}/insights`, {
    method: "POST",
    body,
    accessToken,
  });
}

export async function sendChatMessage(
  body: AiChatRequest,
  _accessToken?: string | null | undefined
): Promise<AiChatResponse> {
  const { sendChatMessageViaApi } = await import("@/services/ai-chat");
  const outcome = await sendChatMessageViaApi(body);
  if (outcome.setupRequired) {
    throw new Error(outcome.setupMessage);
  }
  if (outcome.billingError) {
    throw new Error(outcome.billingMessage);
  }
  if (outcome.error || !outcome.data) {
    throw new Error(outcome.error ?? "Chat failed");
  }
  return outcome.data;
}

export async function fetchChatHistory(
  _accessToken?: string | null | undefined,
  limit = 20
): Promise<ChatHistoryMessage[]> {
  const { fetchChatHistoryViaApi } = await import("@/services/ai-chat");
  const outcome = await fetchChatHistoryViaApi(limit);
  if (outcome.error) return [];
  return outcome.data ?? [];
}
